chain/client/src/test_utils.rs

```diff
@@ -1558,6 +1548,55 @@ impl TestEnv {
     pub fn get_runtime_config(&self, idx: usize, epoch_id: EpochId) -> RuntimeConfig {
         self.clients[idx].runtime_adapter.get_protocol_config(&epoch_id).unwrap().runtime_config
     }
+
+    /// Create and process a SIR tx from actions and return the execution outcome.
+    ///
+    /// The signer defines both sender and receiver of the transaction, hence it
+    /// is a Sender-is-Receiver transaction.
+    pub fn execute_sir_tx(
+        &mut self,
+        actions: Vec<Action>,
+        signer: &InMemorySigner,
+    ) -> FinalExecutionOutcomeView {
+        let tip = self.clients[0].chain.head().unwrap();
+        let tx = Transaction {
+            signer_id: signer.account_id.clone(),
+            receiver_id: signer.account_id.clone(),
+            public_key: signer.public_key(),
+            actions,
+            nonce: tip.height + 1,
+            block_hash: tip.last_block_hash,
+        };
+        let signed_tx = tx.sign(signer);
+        let tx_hash = signed_tx.get_hash().clone();
+        self.clients[0].process_tx(signed_tx, false, false);
+        let max_iters = 1000;
+        for i in 0..max_iters {
+            let block = self.clients[0].produce_block(tip.height + i + 1).unwrap().unwrap();
+            self.process_block(0, block.clone(), Provenance::PRODUCED);
+            if let Ok(outcome) = self.clients[0].chain.get_final_transaction_result(&tx_hash) {
+                return outcome;
+            }
+        }
+        panic!("No transaction outcome found after {max_iters} blocks.")
+    }
+
+    /// Execute a function call transaction that calls main on the `TestEnv`.
+    ///
+    /// This function assumes that account has been deployed and that
+    /// `InMemorySigner::from_seed` produces a valid signer that has it's key
+    /// deployed already.
+    pub fn call_main(&mut self, account: &AccountId) -> FinalExecutionOutcomeView {
+        let signer = InMemorySigner::from_seed(account.clone(), KeyType::ED25519, account.as_str());
+        let actions = vec![Action::FunctionCall(FunctionCallAction {
+            method_name: "main".to_string(),
+            args: vec![],
+            gas: 3 * 10u64.pow(14),
+            deposit: 0,
+        })];
+
+        self.execute_sir_tx(actions, &signer)
+    }
```

Comment:
One thing I don't like here is that the API is non-compistional. If at some point we introduce some kind of transaction builder which returns a Tx, we'd have to mirror the API here, as there isn't a function which takes a Tx. 

This is a non-trivial problem: tx has a nonce, and a block hash, and has to be signed, and we don't want the caller to get the nonce.


So perhaps we need the following, slightly hacky API?

```
pub fn tx(&mut self, mut tx: Transaction) {
  tx.nonce = tip.height + 1;
  tx.block_hash = tip.hash;
  let signer = Signer::from_seed(tx.signer_id.as_str())
  let tx = tx.sign(&signer)
}
```

The caller would then basically specify garbage nonce & hash

Reply:
Hm, I was really hoping to get rid of the caller-site giving garbage nonce and hash.

I have a different suggestion how to make it more compositional, basically split the "create tx" part from the "execute tx" into two functions.

```rust
fn tx_from_actions(
    &mut self,
    actions: Vec<Action>,
    signer: &InMemorySigner,
    receiver: AccountId,
) -> SignedTransaction;

pub fn execute_tx(&mut self, tx: SignedTransaction) -> FinalExecutionOutcomeView;

```

This also opened up more removal of duplicated code in `test_deploy_cost_increased`. The new interface and this additional refactor are included in the latest two commits.

@matklad Do you think this is better now?

Reply:
Yeah, I definitelly like the new API better, in particular, `execute_tx(&mut self, tx: SignedTransaction)` is very much the right boundary. 

I think I'd still prefer garbage nonce approach, but that's debatable. In particular, I feel like that's more of a problem of the split between Tx and SignedTx -- ideally, "what the TX logically is" (list of actions, sender, receiver) should be separate from stuff we need to make cryptograhpy work (nonce, block hash, signature), at least in the testing API. Ideally, we'd have some kind of TxBuilder, but a Tx with `Default` nonce&block hash can stand in its place. But either of these is a bigger change than warranted in this PR, even if they are good ideas.

Reply:
I think the logical piece of work that we generally care about is just a list of actions, not a transaction.

What you describe as 

>  "what the TX logically is" (list of actions, sender, receiver)

in my opinion is not a `Transaction`. The crypto stuff to define valid placement in the chain very much belongs to a valid transaction definition. The difference between `Transaction` and `SignedTransaction` is just how we represent it in memory but logically they describe the same thing.

But I support the idea of a TX builder that prepares actions/signer/receiver and builds into a full TX. But `fn tx_from_actions` is already quite simple to build a TX. I think to make a builder valuable, it would need to do more than just that. For example, track accounts included in genesis, or help somehow with preparing the receiver with a deployed contract first etc.